src/stub_open.c

```diff
@@ -14,21 +15,132 @@
+#define FSYS_ERR_INVALID_PARAMETER      -36 /* (19) Given parameter is invalid */
+
+/*
+File mode mappings.
+https://man7.org/linux/man-pages/man3/fopen.3.html
+http://elm-chan.org/fsw/ff/doc/open.html
+
+│ fopen()     │ open() flags                  │ FatFS
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+│     r       │ O_RDONLY                      │ FA_READ                               │
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+│     w       │ O_WRONLY | O_CREAT | O_TRUNC  │ FA_CREATE_ALWAYS | FA_WRITE           │
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+│     a       │ O_WRONLY | O_CREAT | O_APPEND │ FA_OPEN_APPEND | FA_WRITE             │
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+│     r+      │ O_RDWR                        │ FA_READ | FA_WRITE                    │
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+│     w+      │ O_RDWR | O_CREAT | O_TRUNC    │ FA_CREATE_ALWAYS | FA_WRITE | FA_READ │
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+│     a+      │ O_RDWR | O_CREAT | O_APPEND   | FA_OPEN_APPEND | FA_WRITE | FA_READ   │
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+*/
+static int _map_oflags(int oflag) {
     int result = 0;
-    if (oflag & O_RDONLY) result |= FA_READ;
-    if (oflag & O_WRONLY) result |= FA_WRITE;
-    if (oflag & O_APPEND) result |= FA_OPEN_APPEND;
-    if (oflag & O_CREAT & O_TRUNC) {
-        result |= FA_CREATE_ALWAYS;
-    } else if (oflag & O_CREAT) {
-        result |= FA_CREATE_NEW;
+
+    switch(oflag) {
+        // r
+        case O_RDONLY:
+            result = FA_READ;
+            break;
+        // w
+        case (O_WRONLY | O_CREAT | O_TRUNC):
+            result = FA_CREATE_ALWAYS | FA_WRITE;
+            break;
+        // a
+        case (O_WRONLY | O_CREAT | O_APPEND ):
+            result = FA_OPEN_APPEND | FA_WRITE;
+            break;
+        // r+
+        case (O_RDWR ):
+            result = FA_READ | FA_WRITE;
+            break;
+        // w+
+        case (O_RDWR | O_CREAT | O_TRUNC):
+            result = FA_CREATE_ALWAYS | FA_WRITE | FA_READ;
+            break;
+        // a+
+        case (O_RDWR | O_CREAT | O_APPEND):
+            result = FA_OPEN_APPEND | FA_WRITE | FA_READ;
+            break;
+        default:
+            errno = EINVAL;
+            result = -1;
+            break;
     }
+
     return result;
 }
 
+static void _set_errno(int result) {
+    switch(result) {
+        case FSYS_ERR_DISK_ERR:
+        case FSYS_ERR_INT_ERR:
+        case FSYS_ERR_NOT_READY:
+        case FSYS_ERR_NOT_ENABLED:
+        case FSYS_ERR_NO_FILESYSTEM:
+            errno = ENODEV;
```

Comment:
In all these `errno` settings it should use the `__set_errno()` macro in `errno.h`.